crates/ide_completion/src/context.rs

```diff
@@ -535,3 +628,214 @@ fn path_or_use_tree_qualifier(path: &ast::Path) -> Option<ast::Path> {
     let use_tree = use_tree_list.syntax().parent().and_then(ast::UseTree::cast)?;
     use_tree.path()
 }
+
+#[cfg(test)]
+mod tests {
+    use hir::HirDisplay;
+
+    use crate::test_utils::{position, TEST_CONFIG};
+
+    use super::CompletionContext;
+
+    fn check_expected_type_and_name(
+        ra_fixture: &str,
+        expected_type: Option<&str>,
+        expected_name: Option<&str>,
```

Comment:
That's not too important, but I'd pass `expect: Expect` here, instead of two string arguments. This helps if in the future we decide to add an extra bit of info (perhaps the list of required traits for generic params) or remove something (if name matching doesn't pull its weight).

With expect, you change just the check function itself, and re-run tests to self-update. With specific arguments, you need to update every call-site.

Reply:
Thanks for the suggestion - this is much better!

I'm not sure if there is a standard way used across the codebase for printing optional types. I didn't think the `Some`/`None` strings were very useful, so I just print the value if it is `Some`, or else print `?`. But let me know if there is some other standard I should be following.